File: holding_core/billing.py
from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class PricingPlan(StrEnum):
    FOUNDER = "founder"
    STANDARD = "standard"


class BillingStatus(StrEnum):
    ACTIVE = "active"
    SUBSCRIPTION_REQUIRED = "subscription_required"
    FILING_PACKAGE_REQUIRED = "filing_package_required"
    READY_FOR_PRODUCTION_FILING = "ready_for_production_filing"
    UNSUPPORTED_CASE = "unsupported_case"
    REFUND_ELIGIBLE = "refund_eligible"


class Pricing(BaseModel):
    model_config = ConfigDict(extra="forbid")

    plan: PricingPlan
    monthly_nok: int
    filing_package_nok: int


class CompanyBillingAccount(BaseModel):
    model_config = ConfigDict(extra="forbid")

    company_id: str
    pricing: Pricing
    founder_cohort_number: int | None = Field(default=None, ge=1, le=100)
    subscription_active: bool = False
    filing_package_paid: bool = False
    supported_case: bool = True
    refund_eligible: bool = False


class BillingGateResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: BillingStatus
    allowed: bool
    charge_allowed: bool
    message: str


FOUNDER_PRICING = Pricing(plan=PricingPlan.FOUNDER, monthly_nok=29, filing_package_nok=299)
STANDARD_PRICING = Pricing(plan=PricingPlan.STANDARD, monthly_nok=49, filing_package_nok=499)


def assign_founder_pricing(company_id: str, *, cohort_number: int) -> CompanyBillingAccount:
    return CompanyBillingAccount(
        company_id=company_id,
        pricing=FOUNDER_PRICING,
        founder_cohort_number=cohort_number,
    )


def assign_standard_pricing(company_id: str) -> CompanyBillingAccount:
    return CompanyBillingAccount(company_id=company_id, pricing=STANDARD_PRICING)


def production_filing_gate(account: CompanyBillingAccount, *, filing_ready: bool) -> BillingGateResult:
    if not account.supported_case:
        return BillingGateResult(
            status=BillingStatus.UNSUPPORTED_CASE,
            allowed=False,
            charge_allowed=False,
            message="Case is outside Talli support boundary; do not charge filing package.",
        )
    if not account.subscription_active:
        return BillingGateResult(
            status=BillingStatus.SUBSCRIPTION_REQUIRED,
            allowed=False,
            charge_allowed=False,
            message="Active subscription required before production filing.",
        )
    if not filing_ready:
        return BillingGateResult(
            status=BillingStatus.ACTIVE,
            allowed=False,
            charge_allowed=False,
            message="Filing readiness gate must pass before charging filing package.",
        )
    if not account.filing_package_paid:
        return BillingGateResult(
            status=BillingStatus.FILING_PACKAGE_REQUIRED,
            allowed=False,
            charge_allowed=True,
            message="Filing package payment required before production submission starts.",
        )
    return BillingGateResult(
        status=BillingStatus.READY_FOR_PRODUCTION_FILING,
        allowed=True,
        charge_allowed=False,
        message="Billing and filing readiness gates have passed.",
    )


def mark_supported_filing_failure(account: CompanyBillingAccount) -> CompanyBillingAccount:
    return account.model_copy(update={"refund_eligible": True})

File: app/lib/billing.ts
export type BillingPlan = "founder" | "standard";
export type BillingStatus =
  | "active"
  | "subscription_required"
  | "filing_package_required"
  | "ready_for_production_filing"
  | "unsupported_case"
  | "refund_eligible";

export type BillingAccount = {
  company_id: string;
  pricing_plan: BillingPlan;
  monthly_nok: number;
  filing_package_nok: number;
  founder_cohort_number: number | null;
  subscription_active: boolean;
  filing_package_paid: boolean;
  supported_case: boolean;
  refund_eligible: boolean;
  no_charge_reason: string | null;
};

export type BillingGateResult = {
  status: BillingStatus;
  allowed: boolean;
  chargeAllowed: boolean;
  message: string;
};

export class BillingValidationError extends Error {
  readonly code: string;

  constructor(message: string, code: string) {
    super(message);
    this.name = "BillingValidationError";
    this.code = code;
  }
}

export function billingPricing(plan: BillingPlan) {
  if (plan === "founder") {
    return { pricing_plan: plan, monthly_nok: 29, filing_package_nok: 299 };
  }
  return { pricing_plan: plan, monthly_nok: 49, filing_package_nok: 499 };
}

export function buildBillingAccount(input: {
  companyId: string;
  pricingPlan: BillingPlan;
  founderCohortNumber?: number | null;
  subscriptionActive?: boolean;
  filingPackagePaid?: boolean;
  supportedCase?: boolean;
  refundEligible?: boolean;
  noChargeReason?: string | null;
}): BillingAccount {
  if (!input.companyId) {
    throw new BillingValidationError("Selskap mangler.", "missing_company");
  }
  if (!["founder", "standard"].includes(input.pricingPlan)) {
    throw new BillingValidationError("Ugyldig prisplan.", "invalid_pricing_plan");
  }
  const founderCohortNumber = input.pricingPlan === "founder" ? input.founderCohortNumber : null;
  const validFounderCohort =
    typeof founderCohortNumber === "number" &&
    Number.isInteger(founderCohortNumber) &&
    founderCohortNumber >= 1 &&
    founderCohortNumber <= 100;
  if (input.pricingPlan === "founder" && !validFounderCohort) {
    throw new BillingValidationError("Founder-kull må være mellom 1 og 100.", "founder_cohort_limit");
  }
  const pricing = billingPricing(input.pricingPlan);
  return {
    company_id: input.companyId,
    ...pricing,
    founder_cohort_number: founderCohortNumber ?? null,
    subscription_active: Boolean(input.subscriptionActive),
    filing_package_paid: Boolean(input.filingPackagePaid),
    supported_case: input.supportedCase ?? true,
    refund_eligible: Boolean(input.refundEligible),
    no_charge_reason: input.noChargeReason ?? null,
  };
}

export function productionBillingGate(account: BillingAccount, filingReady: boolean): BillingGateResult {
  if (account.refund_eligible) {
    return {
      status: "refund_eligible",
      allowed: false,
      chargeAllowed: false,
      message: "Filingpakken er markert refusjonsberettiget etter støttet feil.",
    };
  }
  if (!account.supported_case) {
    return {
      status: "unsupported_case",
      allowed: false,
      chargeAllowed: false,
      message: account.no_charge_reason || "Saken er utenfor Talli-støtte. Ikke ta betalt for filingpakke.",
    };
  }
  if (!account.subscription_active) {
    return {
      status: "subscription_required",
      allowed: false,
      chargeAllowed: false,
      message: "Aktivt abonnement kreves før produksjonsfiling.",
    };
  }
  if (!filingReady) {
    return {
      status: "active",
      allowed: false,
      chargeAllowed: false,
      message: "Filing readiness må være klar før filingpakke kan betales.",
    };
  }
  if (!account.filing_package_paid) {
    return {
      status: "filing_package_required",
      allowed: false,
      chargeAllowed: true,
      message: "Filingpakke må betales før produksjonsinnsending.",
    };
  }
  return {
    status: "ready_for_production_filing",
    allowed: true,
    chargeAllowed: false,
    message: "Billing og filing readiness er klare.",
  };
}
